src/style-spec/expression/definitions/formatted.js

```diff
@@ -0,0 +1,145 @@
+// @flow
+
+import { NumberType, ValueType, FormattedType } from '../types';
+
+
+import type { Expression } from '../expression';
+import type EvaluationContext from '../evaluation_context';
+import type ParsingContext from '../parsing_context';
+import type { Type } from '../types';
+
+export class FormattedSection {
+    text: string
+    scale: number | null
+    fontStack: string | null
+
+    constructor(text: string, scale: number | null, fontStack: string | null) {
+        this.text = text;
+        this.scale = scale;
+        this.fontStack = fontStack;
+    }
+}
+
+export class Formatted {
+    sections: Array<FormattedSection>
+
+    constructor(sections: Array<FormattedSection>) {
+        this.sections = sections;
+    }
+
+    toString(): string {
+        return this.sections.map(section => section.text).join('');
+    }
+
+    serialize() {
+        const serialized = ["format"];
+        for (const section of this.sections) {
+            serialized.push(section.text);
+            const fontStack = section.fontStack ?
+                ["literal", section.fontStack.split(',')] :
+                null;
+            serialized.push({ "text-font": fontStack, "font-scale": section.scale });
+        }
+        return serialized;
+    }
+}
+
+type FormattedSectionExpression = {
+    text: Expression,
+    scale: Expression | null;
+    font: Expression | null;
+}
+
+export class FormattedExpression implements Expression {
+    type: Type;
+    sections: Array<FormattedSectionExpression>;
+
+    constructor(sections: Array<FormattedSectionExpression>) {
+        this.type = FormattedType;
+        this.sections = sections;
+    }
+
+    static parse(args: Array<mixed>, context: ParsingContext): ?Expression {
+        if (args.length < 3) {
+            return context.error(`Expected at least two arguments.`);
+        }
+
+        if ((args.length - 1) % 2 !== 0) {
+            return context.error(`Expected an even number of arguments.`);
+        }
+
+        const sections: Array<FormattedSectionExpression> = [];
+        for (let i = 1; i < args.length - 1; i += 2) {
+            const text = context.parse(args[i], 1, ValueType);
+            if (!text) return null;
+            const kind = text.type.kind;
+            if (kind !== 'string' && kind !== 'value' && kind !== 'null')
+                return context.error(`Formatted text type must be 'string', 'value', or 'null'.`);
```

Comment:
Basically support anything that has a string coercion? I'm ambivalent here, but I'll follow your guidance since you've been so involved in the explicit vs. implicit coercion design choices.

My starting position would be:
 - It's easier to reason about the implementation without automatic coercion
 - I can't immediately think of a "gotcha" case where this would be super unergonomic. If you want to format a plain number, you'll get an error message, but it'll be pretty clear what the error message means and there's a straightforward coercion call to address it.
 - We can always relax the requirements in the future